File: day05/aoc2025_05.scala
object aoc2025_05 {
    def main(args: Array[String]): Unit = {
        val lines = scala.io.Source.fromFile("day05/input05").getLines().toArray
        val (intervals: Array[Array[Long]], nums: Array[Long]) = 
        lines.splitAt(lines.indexOf("")) match {
            case (a, b) => (a.map(_.split("-").map(_.toLong)), 
            b.tail.map(_.toLong))
        }
        
        val num_contained = nums.count(n => intervals.exists(arr => arr(0) <= n && n <= arr(1)))
        println("Part 1: " + num_contained)
        
        intervals.sortInPlaceBy(_(0))
        for (i <- Range(1, intervals.length)) {
            if (intervals(i)(0) <= intervals(i-1)(1)) {
                intervals(i)(0) = intervals(i-1)(0)
                intervals(i)(1) =  intervals(i)(1).max(intervals(i-1)(1))
                intervals(i-1) = null
            }
        }
        println("Part 2: " + 
        intervals.map(x => if (x == null) 0 else x(1) - x(0) + 1).sum)
    }
}

File: day05/aoc2025_05.ts
import { readFileSync } from "node:fs";

const lines = readFileSync("day05/input05", "utf-8").trim().split("\n")
const separator = lines.indexOf('')
const intervals: number[][] = lines.slice(0, separator)
    .map(s => s.split('-').map(Number))
const nums: number[] = lines.slice(separator + 1).map(Number)

const num_contained = nums
    .filter(n => intervals.some(arr => arr[0] <= n && n <= arr[1]))
    .length
console.log('Part 1:', num_contained)

intervals.sort((a, b) => a[0] - b[0])
for (let i = 1; i < intervals.length; i++) {
    if (intervals[i][0] <= intervals[i - 1][1]) {
        intervals[i][0] = intervals[i - 1][0]
        intervals[i][1] = Math.max(intervals[i][1], intervals[i - 1][1])
        intervals[i - 1] = []
    }
}
const part2 = intervals
    .map(x => x.length === 0 ? 0 : x[1] - x[0] + 1)
    .reduce((x, y) => x + y)
console.log('Part 2:', part2)
